plugins/system_upgrade.py

```diff
@@ -568,6 +576,8 @@ def run_reboot(self):
         if not self.opts.tid[0] == "reboot":
             return
 
+        self.state.poweroff_after = self.opts.poweroff_after
+
         self.log_status(_("Rebooting to perform upgrade."),
                         REBOOT_REQUESTED_ID)
         reboot()
```

Comment:
I was testing the functionality today and it seems there is a problem with serializing the `State` object inside the `__exit__` method, because it is not executed when triggering the reboot using the `Popen`.

As this is the only place where we want to serialize it explicitly, we can call the `self.state.write()` on this line. I was testing that and it seems working.

Reply:
Yep, added that line and it builds now, thanks. Also added a check before writing because the unit tests were trying to write with the MagicMock object.

Reply:
Hi and thanks for testing this! I would prefer not adding any test-related workarounds inside the execution code. We can avoid the error in the unit tests simply by putting the `self.command.opts.poweroff_after = False` inside the `test_run_reboot` test. Could you please update the PR?

Also, probably the last thing, could you add a comment to the `self.state.write()` line, to know why we are doing that at that place. Something like that we need to save the state explicitly here, because the `__exit__` hook is not called when triggering the reboot. Thanks!

Reply:
Updated and pushed